homer0/projext-plugin-webpack-react
# Update the plugin

### What does this PR do?

I rewrote the whole thing:

**`includeTargets` > `frameworkOptions.ssr`**

Before this, if you wanted to implement SSR, you would have to go to your Node target settings and set `framework` to `react`, and `frameworkOptions.src` with the name of the frontend target you wanted to serve.

Well... that's gone. Thanks to the new [`includeTargets` setting](https://github.com/homer0/projext/pull/41), projext can take care of including the paths for other targets while defining the new [file rules](https://github.com/homer0/projext/pull/40).

**Babel Helper**

[Babel helper](https://github.com/homer0/projext/pull/38) is a new projext service that allows you to update a Babel configuration using really simple methods. Thanks to this, adding the required preset and plugin was reduced to two simple calls.

**Externals**

This was a bug. On the current version, if you implement SSR on a Node target, you have to go to the target settings and add `react-dom/server` and `react` to `excludeModules`, so webpack won't try to bundle them in your build.

Well, now the plugin detects if your target type is Node or if its a library and automatically adds them as external dependencies.

**EVERYTHING is _"private/protected"_ now**

I'm doing this on several plugins and projext services now: Methods and properties that weren't intended to be used outside its class are now prefixed with an underscore (`_`), to mark them as _"protected/private"_.

Sometimes I consider some methods/properties to be "ok" as _"public"_ as they may be of utility for someone, but it wasn't until a few releases ago that I realized that by thinking that, changing those methods/properties, would require a breaking release, when there was no "real" reason for those things to be _"public"_.

So, from now on, use events and reducers and try to stay away from methods and properties prefixed with underscore.

**Dependencies up to date**

All production and development dependencies were updated to their latest version.

Now, there's a special mention: ESLint. It latest version is `v5`, but the Airbnb preset I use is still not compatible with it, so I updated ESLint to the last version before `v5`: `v4.19.1`.

I also updated a few things on the `.eslintrc` after getting the Airbnb preset last version:

- `lines-between-class-members`: I disabled this rule because I already fill the space between methods with comments, so I see not reason for it.
- `operator-linebreak`: I changed the default value from the Airbnb, `before`,  to `after`. The convention I use and which I believe is easier to read is with the operators at the end of the line.

**`hooks` instead of `install-hooks`**

The `yarn`/`npm` script to install the repositories hook has been changed to just `hooks`.

### How should it be tested manually?

> You need to use `homer0/projext#next` as version for `projext` and `homer0/projext-plugin-webpack#next` for `projext-plugin-webpack`.

First,  the _"real breaking change"_ is that `frameworkOptions.ssr` doesn't do anything now, so you should try replacing it with `includeTargets` and make sure your SSR app still works.

Then, you can try removing `react` and `react-dom/server` from `excludeModules`, since it's already provided by the plugin, webpack shouldn't try to bundle them.

And of course...

```bash
npm test
# or
yarn test
```
